src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt

```diff
@@ -0,0 +1,62 @@
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.search.LocalSearchScope
+import com.intellij.psi.search.searches.ReferencesSearch
+import org.rust.ide.annotator.fixes.RemoveMutableFix
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.isMut
+import org.rust.lang.core.psi.ext.parentOfType
+import org.rust.lang.core.psi.ext.selfParameter
+import org.rust.lang.core.resolve.ref.RsMethodCallReferenceImpl
+
+class RsVariableMutableInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "No mutable required"
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) =
+        object : RsVisitor() {
+            override fun visitPatBinding(o: RsPatBinding) {
+                if (!o.isMut) return
+                val block = o.parentOfType<RsBlock>() ?: o.parentOfType<RsFunction>() ?: return
+                val occurrences = ReferencesSearch.search(o, LocalSearchScope(block))
+                    .map { it.element.parent }
+                var isMutableNeeded = false
+                if (!occurrences.isEmpty()) {
+                    loop@ for (occurrence in occurrences) {
+                        val parent = occurrence.parent
+                        when (parent) {
+                            is RsUnaryExpr -> {
+                                if (parent.mut != null) {
```

Comment:
Unnecessary nested blocks here and below: `is RsUnaryExpr -> if (...) {` will work. And overall the code is hard to analyse because of many nested levels. It would be useful to break it into small and focused functions.